Sources/SwiftDocC/Infrastructure/DocumentationContext.swift

```diff
@@ -2328,34 +2333,81 @@ public class DocumentationContext: DocumentationContextDataProviderDelegate {
         return automaticallyCuratedSymbols
     }
 
-    /// Handles overloaded symbols by grouping them together into one page.
-    private func groupOverloadedSymbols(with linkResolver: PathHierarchyBasedLinkResolver) throws {
+    private func addOverloadGroupReferences(overloadGroups: [String: [String]]) throws {
         guard FeatureFlags.current.isExperimentalOverloadedSymbolPresentationEnabled else {
             return
         }
         
-        try linkResolver.traverseOverloadedSymbols { overloadedSymbolReferences in
+        for (overloadGroupId, var overloadSymbolIds) in overloadGroups {
+            guard overloadSymbolIds.count > 1 else {
+                assertionFailure("SymbolKit should only create overload groups of more than one symbol")
+                continue
+            }
+            guard let overloadGroupReference = documentationCache.reference(symbolID: overloadGroupId) else {
+                preconditionFailure("Overload group symbol should already be in the cache")
+            }
+            let overloadGroupNode = try entity(with: overloadGroupReference)
 
-            // Tell each symbol what other symbols overload it.
-            for (index, symbolReference) in overloadedSymbolReferences.indexed() {
-                let documentationNode = try entity(with: symbolReference)
+            var overloadSymbolNodes = try overloadSymbolIds.map {
+                guard let reference = documentationCache.reference(symbolID: $0) else {
+                    preconditionFailure("Symbols should already be in the cache")
+                }
+                return try entity(with: reference)
+            }
+            if overloadSymbolNodes.allSatisfy({ node in
+                (node.semantic as? Symbol)?.overloadsVariants.firstValue != nil
+            }) {
+                // If SymbolKit saved its sort of the symbols, use that ordering here
+                overloadSymbolNodes.sort(by: { lhs, rhs in
+                    let lhsIndex = (lhs.semantic as! Symbol).overloadsVariants.firstValue!.displayIndex
+                    let rhsIndex = (rhs.semantic as! Symbol).overloadsVariants.firstValue!.displayIndex
+                    return lhsIndex < rhsIndex
+                })
+            } else {
+                assertionFailure("Failed to find the overload data from SymbolKit.")
```

Comment:
Can we add more information to this assertion message to help debug issues we ever hit this?